FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/HandlePasswordProtectedMsgWorker.kt

```diff
@@ -0,0 +1,425 @@
+
+      var list: List<MessageEntity>
+      var lastMsgUID = 0L
+      while (true) {
+        list = roomDatabase.msgDao().getOutboxMsgsByStatesSuspend(
+          account = account.email,
+          msgStates = listOf(MessageState.NEW_PASSWORD_PROTECTED.value)
+        )
+
+        if (list.isEmpty()) {
+          break
+        }
+
+        val iterator = list.iterator()
+        var msgEntity: MessageEntity? = null
+
+        while (iterator.hasNext()) {
+          val tempMsgDetails = iterator.next()
+          if (tempMsgDetails.uid > lastMsgUID) {
+            msgEntity = tempMsgDetails
+            break
+          }
+        }
+
+        if (msgEntity == null) {
+          msgEntity = list[0]
+        }
+
+        lastMsgUID = msgEntity.uid
+
+        try {
+          if (msgEntity.isEncrypted == true && msgEntity.isPasswordProtected) {
+            //get msg attachments(including forwarded that can be encrypted)
+            val attachments = roomDatabase.attachmentDao().getAttachmentsSuspend(
+              account = msgEntity.email,
+              label = JavaEmailConstants.FOLDER_OUTBOX,
+              uid = msgEntity.uid
+            ).map {
+              it.copy(
+                forwardedFolder = "Outbox",
+                forwardedUid = Random.nextLong(),
+                decryptWhenForward = true
+              )
+            }
+
+            //create MimeMessage from content + attachments
+            val mimeMsgWithAttachments = EmailUtil.createMimeMsg(
+              context = applicationContext,
+              sess = Session.getDefaultInstance(Properties()),
+              account = account,
+              msgEntity = msgEntity.copy(isEncrypted = false),
+              atts = attachments
+            )
+
+            //get recipients that will be used to create password-encrypted msg
+            val toCandidates = mimeMsgWithAttachments.getMatchingRecipients(
+              type = Message.RecipientType.TO,
+              list = GeneralUtil.getRecipientsWithoutUsablePubKeys(
+                context = applicationContext,
+                emails = mimeMsgWithAttachments.getAddresses(Message.RecipientType.TO)
+              )
+            )
+
+            val ccCandidates = mimeMsgWithAttachments.getMatchingRecipients(
+              type = Message.RecipientType.CC,
+              list = GeneralUtil.getRecipientsWithoutUsablePubKeys(
+                context = applicationContext,
+                emails = mimeMsgWithAttachments.getAddresses(Message.RecipientType.CC)
+              )
+            )
+
+            val bccCandidates = mimeMsgWithAttachments.getMatchingRecipients(
+              type = Message.RecipientType.BCC,
+              list = GeneralUtil.getRecipientsWithoutUsablePubKeys(
+                context = applicationContext,
+                emails = mimeMsgWithAttachments.getAddresses(Message.RecipientType.BCC)
+              )
+            )
```

Comment:
Please extract these three into a separate method. Overall, there is too much code here in one method and a lot of it could be extracted, which will make the code easier to follow.

Reply:
will do